File: ts/src/gate.ts
/**
 * 렌더前 invariant gate (REEL-QA-GATE)
 *
 * seam 통일규칙 #3: 음성=영상=타임라인 정렬, 전 트랙 믹스, 자막 커버리지를
 * *증명*해야 렌더. 미달 = block (false-DONE 구조차단 + 음소거 슬라이드쇼 재발 방지).
 *
 * Atropos가 CompositionSnapshot 만들기 전에 호출. 통과 못하면 렌더 금지.
 */

import { BeatPlan } from './beat-plan';
import { AudioClip } from './audio-clip';
import { MediaArtifact } from './media-artifact';

export interface RenderReadiness {
  ok: boolean;
  violations: string[];
  warnings: string[];
}

/**
 * assert_render_ready — 전 비트가 보이스(P1)·자막(P13)·비주얼을 갖췄는지 증명
 *
 * ★ ENHANCED BEAT COVERAGE: beat_index를 통해 모든 audio/media 클립 커버리지 검증
 * - voice/sfx AudioClip이 beat_index를 선언해야 하므로 orphan clip 자동 감지
 * - MediaArtifact도 beat_index 필수이므로 누락된 시각 자동 감지
 * - 결과: "어제 음소거" 버그 구조적 불가능화
 */
export function assertRenderReady(
  plan: BeatPlan,
  audioList: AudioClip[],
  mediaList: MediaArtifact[],
  options: {
    requireVoicePerBeat?: boolean;
    requireCaptionPerBeat?: boolean;
  } = {}
): RenderReadiness {
  const requireVoicePerBeat = options.requireVoicePerBeat !== false;
  const requireCaptionPerBeat = options.requireCaptionPerBeat !== false;

  const violations: string[] = [];
  const warnings: string[] = [];

  const beatIndices = new Set(plan.beats.map(b => b.beat_index));
  if (beatIndices.size === 0) {
    return {
      ok: false,
      violations: ['BeatPlan에 비트 0개'],
      warnings: [],
    };
  }

  // 개별 계약 자체 위반 먼저 (P1 결박)
  for (const clip of audioList) {
    if (clip.track === 'voice' || clip.track === 'sfx') {
      if (clip.beat_index === undefined || clip.beat_index === null) {
        violations.push(
          `P1 위반: audio ${clip.track} beat_index=null (orphan 클립, 침묵 위험)`
        );
      }
    }
  }

  // MediaArtifact beat_index 커버리지
  for (const mediaItem of mediaList) {
    if (mediaItem.beat_index === undefined || mediaItem.beat_index === null) {
      violations.push(`media beat_index=null (orphan 클립)`);
    }
  }

  // 보이스 커버리지 (P1 — 보이스 미발화)
  const voiceBeatIndices = new Set(
    audioList
      .filter(c => c.track === 'voice' && c.beat_index !== undefined && c.beat_index !== null)
      .map(c => c.beat_index as number)
  );

  if (requireVoicePerBeat) {
    const missingVoice = Array.from(beatIndices)
      .filter(b => !voiceBeatIndices.has(b))
      .sort((a, b) => a - b);
    if (missingVoice.length > 0) {
      violations.push(`P1 보이스 없는 비트 ${JSON.stringify(missingVoice)} (음소거 위험)`);
    }
  }

  // 비주얼 커버리지
  const mediaBeatIndicesSet = new Set(
    mediaList
      .filter(m => m.beat_index !== undefined && m.beat_index !== null)
      .map(m => m.beat_index as number)
  );

  const missingMedia = Array.from(beatIndices)
    .filter(b => !mediaBeatIndicesSet.has(b))
    .sort((a, b) => a - b);
  if (missingMedia.length > 0) {
    violations.push(`비주얼 없는 비트 ${JSON.stringify(missingMedia)}`);
  }

  // P13 — 자막 커버리지
  const captionBeatIndices = new Set(
    plan.beats
      .filter(b => b.caption && b.caption.trim().length > 0)
      .map(b => b.beat_index)
  );

  if (requireCaptionPerBeat) {
    const missingCaption = Array.from(beatIndices)
      .filter(b => !captionBeatIndices.has(b))
      .sort((a, b) => a - b);
    if (missingCaption.length > 0) {
      warnings.push(`P13 자막 없는 비트 ${JSON.stringify(missingCaption)} (dead air 위험)`);
    }
  }

  // 음악 트랙 확인
  const hasMusic = audioList.some(c => c.track === 'music');
  if (!hasMusic) {
    warnings.push('음악 트랙 없음');
  }

  return {
    ok: violations.length === 0,
    violations,
    warnings,
  };
}

File: hiob_contracts/gate.py
"""렌더前 invariant gate (REEL-QA-GATE) — 계약 레벨에서 어젯밤 6결손을 차단.

seam 통일규칙 #3: 음성=영상=타임라인 정렬, 전 트랙 믹스, 자막 커버리지를
*증명*해야 렌더. 미달 = block (false-DONE 구조차단 + 음소거 슬라이드쇼 재발 방지).

Atropos가 CompositionSnapshot 만들기 전에 호출. 통과 못하면 렌더 금지.
"""
from __future__ import annotations

from dataclasses import dataclass, field

from .beat_plan import BeatPlan
from .audio_clip import AudioClip
from .media_artifact import MediaArtifact


@dataclass(frozen=True)
class RenderReadiness:
    ok: bool
    violations: tuple[str, ...] = ()
    warnings: tuple[str, ...] = ()


def assert_render_ready(
    plan: BeatPlan,
    audio: list[AudioClip],
    media: list[MediaArtifact],
    *,
    require_voice_per_beat: bool = True,
    require_caption_per_beat: bool = True,
) -> RenderReadiness:
    """전 비트가 보이스(P1)·자막(P13)·비주얼을 갖췄는지 증명. 미달=block."""
    violations: list[str] = []
    warnings: list[str] = []

    beat_idxs = {b.beat_index for b in plan.beats}
    if not beat_idxs:
        return RenderReadiness(ok=False, violations=("BeatPlan에 비트 0개",))

    voice_beats = {c.beat_index for c in audio if c.track == "voice" and c.beat_index is not None}
    media_beats = {m.beat_index for m in media if m.beat_index is not None}
    caption_beats = {b.beat_index for b in plan.beats if (b.caption or "").strip()}
    has_music = any(c.track == "music" for c in audio)

    # 계약 자체 위반 먼저(P1 결박)
    for c in audio:
        violations.extend(f"audio {c.track}@{c.beat_index}: {e}" for e in c.validate())
    for m in media:
        violations.extend(f"media @{m.beat_index}: {e}" for e in m.validate())

    # P1 — 보이스 미발화(어젯밤 #1)
    if require_voice_per_beat:
        missing_voice = sorted(beat_idxs - voice_beats)
        if missing_voice:
            violations.append(f"P1 보이스 없는 비트 {missing_voice} (음소거 위험)")

    # 비주얼 커버리지
    missing_media = sorted(beat_idxs - media_beats)
    if missing_media:
        violations.append(f"비주얼 없는 비트 {missing_media}")

    # P13 — 자막 커버리지(없으면 dead air)
    if require_caption_per_beat:
        missing_cap = sorted(beat_idxs - caption_beats)
        if missing_cap:
            warnings.append(f"P13 자막 없는 비트 {missing_cap} (dead air 위험)")

    if not has_music:
        warnings.append("음악 트랙 없음")

    return RenderReadiness(
        ok=not violations,
        violations=tuple(violations),
        warnings=tuple(warnings),
    )
